lib/IDE/CodeCompletion.cpp

```diff
@@ -939,10 +939,15 @@ static CodeCompletionResult::ExpectedTypeRelation calculateTypeRelation(
       Ty->is<ErrorType>() ||
       ExpectedTy->is<ErrorType>())
     return CodeCompletionResult::ExpectedTypeRelation::Unrelated;
-  if (Ty->isEqual(ExpectedTy))
-    return CodeCompletionResult::ExpectedTypeRelation::Identical;
-  if (isConvertibleTo(Ty, ExpectedTy, *DC))
-    return CodeCompletionResult::ExpectedTypeRelation::Convertible;
+
+  // Equality/Conversion of GenericTypeParameterType won't account for
+  // requirements – ignore them
+  if (!Ty->hasTypeParameter() && !ExpectedTy->hasTypeParameter()) {
+    if (Ty->isEqual(ExpectedTy))
+      return CodeCompletionResult::ExpectedTypeRelation::Identical;
+    if (isConvertibleTo(Ty, ExpectedTy, *DC))
```

Comment:
If you look at isConvertibleTo it has some flags to turn type parameters or archetypes into type variables - this will produce a more accurate result. Might just be a matter of reusing the existing mechanism

Reply:
I tried changing isConvertibleTo to canPossiblyConvertTo (which passes openArchetypes=true) but does that make sense here since the two type parameters can have different generic contexts?

E.g. with the below snippet:
```
func foo<U,V>(_ _: U, _ _: V) where U: Sequence, V: Sequence {}
func bar<T, U>(lhs: T, rhs: U) where T: Collection {
  _ = foo(lhs, #^COMPLETE^#
}
```
when adding type relations to the completion items for `lhs` we end up comparing Ty = T to ExpectedTy = V and isEqual and canPossiblyConvertTo both return true. For `rhs` we compare Ty = U to ExpectedTy = V where isEqual is false and canPossiblyConvertTo is true.